### ISSUE TITLE
Feature: fuzzy matching for completion

### ISSUE THREAD

**Issue by PythonNut** [davidhalter/jedi#827]
By "fuzzy matching" I mean Sublime Text style `tifn` -> <code>**t**his\_**i**s\_**a**\_**f**unction\_**n**ame</code>.

This comes from these issues:

* https://github.com/PythonNut/company-flx/issues/17 (fuzzy completion package for Emacs w/ `company-mode`)
* https://github.com/proofit404/company-anaconda/issues/24 (python completion package for Emacs w/ `company-mode`)

I spent some time reading the [`anaconda-mode` source](https://github.com/proofit404/anaconda-mode/blob/4f84759cab7746cf705f75719e701551d47de1e3/anaconda_mode.py#L61), and it appears that support for fuzzy matching needs to be implemented in `jedi`.

How would this work? My uneducated guess is that `jedi.Script.completions` would take an optional parameter to specify whether to return fuzzy completions.

I realize that this could be a very difficult change to make, and that I can't anticipate potential performance implications, but I'd very much like to see this happen. 😄 

_Note: as for me just submitting a PR, it's possible, but I don't think I can for a very long time, if ever._

**Comment by proofit404** [comment]
As far as I understand, company-flx need list of completion candidates and provide fuzzy search on existing list. Jedi on its side provide completion list regarding to code source context. For example, method list of some class doesn't change if you want select from it in fuzzy mode or not. I think this issue relate to anaconda-mode only.

**Comment by PythonNut** [comment]
@proofit404 if we can get `jedi` to return every potential candidate at a given point (ignoring any partial symbol before the point), then we can do everything in `anaconda-mode`.

**Comment by proofit404** [comment]
I guess best solution here will be preprocess buffer on anaconda-mode side before submit it to jedi.

**Comment by PythonNut** [comment]
@proofit404 if that could be done consistently in a safe manner, I suppose that would work as well. Is it really as simple as deleting the symbol before the point?

**Comment by davidhalter** [comment]
I would like to see this.

**Comment by proofit404** [comment]
I misunderstand what way we should do this?

**Comment by davidhalter** [comment]
@proofit404 Sorry I haven't read the whole conversation. I have just read the first post. It was meant as I would like to see this feature without a solution attache to it.

@PythonNut You don't even have to remove the symbol before the point. It's basically something that can be done in `jedi.api.completion` (dev branch!). It's a very small change that doesn't need a lot of work. Note that if you're on master branch it probably doesn't look as simple, because I have almost completely rewritten that stuff.

**Comment by PythonNut** [comment]
@davidhalter that sounds great! I guess we'll wait until that branch gets merged and go from there?

**Comment by davidhalter** [comment]
@PythonNut The branch was already merged. You're free to add this.

**Comment by doolio** [comment]
Any progress on this? It would be most welcome.

**Comment by jmfrank63** [comment]
@davidhalter Hi David,

I finally found the time to continue on this. I got some working code, however with the latest master my simple tests that used to work are now failing (without my changes).

```
import jedi
print(jedi.Script('import os\nos.').completions())

Traceback (most recent call last):
  File "/Users/johannes/Documents/Projects/Python/jedi/jedi/inference/filters.py", line 62, in _get_definition_names
    return for_module[name_key]
KeyError: 'os'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "ctest.py", line 3, in <module>
    print(jedi.Script('import os\nos.').completions())
  File "/Users/johannes/Documents/Projects/Python/jedi/jedi/api/__init__.py", line 217, in completions
    return completion.completions()
  File "/Users/johannes/Documents/Projects/Python/jedi/jedi/api/completion.py", line 99, in completions
    completion_names = self._get_value_completions(leaf)
  File "/Users/johannes/Documents/Projects/Python/jedi/jedi/api/completion.py", line 208, in _get_value_completions
    completion_names += self._trailer_completions(dot.get_previous_leaf())
  File "/Users/johannes/Documents/Projects/Python/jedi/jedi/api/completion.py", line 241, in _trailer_completions
    values = infer_call_of_leaf(inferred_context, previous_leaf)
  File "/Users/johannes/Documents/Projects/Python/jedi/jedi/inference/helpers.py", line 77, in infer_call_of_leaf
    return context.infer_node(leaf)
  File "/Users/johannes/Documents/Projects/Python/jedi/jedi/inference/context.py", line 210, in infer_node
    return infer_node(self, node)
  File "/Users/johannes/Documents/Projects/Python/jedi/jedi/inference/syntax_tree.py", line 139, in infer_node
    return _infer_node_if_inferred(context, element)
  File "/Users/johannes/Documents/Projects/Python/jedi/jedi/inference/syntax_tree.py", line 152, in _infer_node_if_inferred
    return _infer_node_cached(context, element)
  File "/Users/johannes/Documents/Projects/Python/jedi/jedi/inference/cache.py", line 42, in wrapper
    rv = function(obj, *args, **kwargs)
  File "/Users/johannes/Documents/Projects/Python/jedi/jedi/inference/syntax_tree.py", line 157, in _infer_node_cached
    return _infer_node(context, element)
  File "/Users/johannes/Documents/Projects/Python/jedi/jedi/debug.py", line 82, in wrapper
    return func(*args, **kwargs)
  File "/Users/johannes/Documents/Projects/Python/jedi/jedi/inference/syntax_tree.py", line 53, in wrapper
    return func(context, *args, **kwargs)
  File "/Users/johannes/Documents/Projects/Python/jedi/jedi/inference/syntax_tree.py", line 167, in _infer_node
    return infer_atom(context, element)
  File "/Users/johannes/Documents/Projects/Python/jedi/jedi/inference/syntax_tree.py", line 292, in infer_atom
    return context.py__getattribute__(atom, position=position)
  File "/Users/johannes/Documents/Projects/Python/jedi/jedi/inference/context.py", line 43, in py__getattribute__
    names = self.goto(name_or_str, position)
  File "/Users/johannes/Documents/Projects/Python/jedi/jedi/inference/context.py", line 32, in goto
    names = finder.filter_name(filters, name_or_str)
  File "/Users/johannes/Documents/Projects/Python/jedi/jedi/inference/finder.py", line 37, in filter_name
    names = filter.get(string_name)
  File "/Users/johannes/Documents/Projects/Python/jedi/jedi/inference/filters.py", line 240, in get
    return [n for filter in self._filters for n in filter.get(name)]
  File "/Users/johannes/Documents/Projects/Python/jedi/jedi/inference/filters.py", line 240, in <listcomp>
    return [n for filter in self._filters for n in filter.get(name)]
  File "/Users/johannes/Documents/Projects/Python/jedi/jedi/inference/filters.py", line 82, in get
    _get_definition_names(self._used_names, name),
  File "/Users/johannes/Documents/Projects/Python/jedi/jedi/inference/filters.py", line 66, in _get_definition_names
    name for name in names if name.is_definition(include_setitem=True)
  File "/Users/johannes/Documents/Projects/Python/jedi/jedi/inference/filters.py", line 66, in <genexpr>
    name for name in names if name.is_definition(include_setitem=True)
TypeError: is_definition() got an unexpected keyword argument 'include_setitem'
```
I can look into this, but I thought you will probably know exactly what this is.

The fuzzy changes are really small, just testing is a real challenge ;-) Neither VSCode nor NeoVim seem to like taking jedi from a local repo that is using  `pip install -e .`

**Comment by davidhalter** [comment]
Fixed in  #1409 by @jmfrank63
